File: Python/OneLang/One/Transforms/InferTypesPlugins/ResolveMethodCalls.py
from onelang_core import *
import OneLang.One.Transforms.InferTypesPlugins.Helpers.InferTypesPlugin as inferTypesPlug
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Transforms.InferTypesPlugins.Helpers.GenericsResolver as genRes
import OneLang.One.Ast.References as refs
import OneLang.One.Ast.Types as types

class ResolveMethodCalls(inferTypesPlug.InferTypesPlugin):
    def __init__(self):
        super().__init__("ResolveMethodCalls")
    
    def find_method(self, cls_, method_name, is_static, args):
        all_bases = list(filter(lambda x: isinstance(x, types.Class), cls_.get_all_base_interfaces())) if isinstance(cls_, types.Class) else cls_.get_all_base_interfaces()
        
        methods = []
        for base in all_bases:
            for m in base.methods:
                min_len = len(list(filter(lambda p: p.initializer == None, m.parameters)))
                max_len = len(m.parameters)
                match = m.name == method_name and m.is_static == is_static and min_len <= len(args) and len(args) <= max_len
                if match:
                    methods.append(m)
        
        if len(methods) == 0:
            raise Error(f'''Method \'{method_name}\' was not found on type \'{cls_.name}\' with {len(args)} arguments''')
        elif len(methods) > 1:
            # TODO: actually we should implement proper method shadowing here...
            this_methods = list(filter(lambda x: x.parent_interface == cls_, methods))
            if len(this_methods) == 1:
                return this_methods[0]
            raise Error(f'''Multiple methods found with name \'{method_name}\' and {len(args)} arguments on type \'{cls_.name}\'''')
        return methods[0]
    
    def resolve_return_type(self, expr, generics_resolver):
        generics_resolver.collect_from_method_call(expr)
        
        i = 0
        
        while i < len(expr.args):
            # actually doesn't have to resolve, but must check if generic type confirm the previous argument with the same generic type
            param_type = generics_resolver.resolve_type(expr.method.parameters[i].type, False)
            if param_type != None:
                expr.args[i].set_expected_type(param_type)
            expr.args[i] = self.main.run_plugins_on(expr.args[i])
            generics_resolver.collect_resolutions_from_actual_type(param_type, expr.args[i].actual_type)
            i = i + 1
        
        if expr.method.returns == None:
            self.error_man.throw(f'''Method ({expr.method.parent_interface.name}::{expr.method.name}) return type was not specified or infered before the call.''')
            return
        
        expr.set_actual_type(generics_resolver.resolve_type(expr.method.returns, True), True, isinstance(expr, exprs.InstanceMethodCallExpression) and astTypes.TypeHelper.is_generic(expr.object.get_type()))
    
    def transform_method_call(self, expr):
        if isinstance(expr.object, refs.ClassReference) or isinstance(expr.object, refs.StaticThisReference):
            cls_ = expr.object.decl if isinstance(expr.object, refs.ClassReference) else expr.object.cls_ if isinstance(expr.object, refs.StaticThisReference) else None
            method = self.find_method(cls_, expr.method_name, True, expr.args)
            result = exprs.StaticMethodCallExpression(method, expr.type_args, expr.args, isinstance(expr.object, refs.StaticThisReference))
            self.resolve_return_type(result, genRes.GenericsResolver())
            return result
        else:
            resolved_object = expr.object if expr.object.actual_type != None else self.main.run_plugins_on(expr.object)
            object_type = resolved_object.get_type()
            intf_type = object_type.decl if isinstance(object_type, astTypes.ClassType) else object_type.decl if isinstance(object_type, astTypes.InterfaceType) else None
            
            if intf_type != None:
                method = self.find_method(intf_type, expr.method_name, False, expr.args)
                result = exprs.InstanceMethodCallExpression(resolved_object, method, expr.type_args, expr.args)
                self.resolve_return_type(result, genRes.GenericsResolver.from_object(resolved_object))
                return result
            elif isinstance(object_type, astTypes.AnyType):
                expr.set_actual_type(astTypes.AnyType.instance)
                return expr
            else:
                pass
            return resolved_object
    
    def can_transform(self, expr):
        return isinstance(expr, exprs.UnresolvedMethodCallExpression) and not (isinstance(expr.actual_type, astTypes.AnyType))
    
    def transform(self, expr):
        return self.transform_method_call(expr)

File: PHP/src/OneLang/One/Transforms/InferTypesPlugins/ResolveMethodCalls.php
<?php

namespace OneLang\One\Transforms\InferTypesPlugins\ResolveMethodCalls;

use OneLang\One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\UnresolvedMethodCallExpression;
use OneLang\One\Ast\Expressions\InstanceMethodCallExpression;
use OneLang\One\Ast\Expressions\StaticMethodCallExpression;
use OneLang\One\Ast\Expressions\IMethodCallExpression;
use OneLang\One\Ast\Expressions\LambdaCallExpression;
use OneLang\One\Ast\AstTypes\ClassType;
use OneLang\One\Ast\AstTypes\InterfaceType;
use OneLang\One\Ast\AstTypes\AnyType;
use OneLang\One\Ast\AstTypes\TypeHelper;
use OneLang\One\Ast\AstTypes\LambdaType;
use OneLang\One\Transforms\InferTypesPlugins\Helpers\GenericsResolver\GenericsResolver;
use OneLang\One\Ast\References\ClassReference;
use OneLang\One\Ast\References\InstanceFieldReference;
use OneLang\One\Ast\References\StaticThisReference;
use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\Types\IInterface;
use OneLang\One\Ast\Types\Method;

class ResolveMethodCalls extends InferTypesPlugin {
    function __construct() {
        parent::__construct("ResolveMethodCalls");
        
    }
    
    protected function findMethod($cls, $methodName, $isStatic, $args) {
        $allBases = $cls instanceof Class_ ? array_values(array_filter($cls->getAllBaseInterfaces(), function ($x) { return $x instanceof Class_; })) : $cls->getAllBaseInterfaces();
        
        $methods = array();
        foreach ($allBases as $base)
            foreach ($base->methods as $m) {
                $minLen = count(array_values(array_filter($m->parameters, function ($p) { return $p->initializer === null; })));
                $maxLen = count($m->parameters);
                $match = $m->name === $methodName && $m->isStatic === $isStatic && $minLen <= count($args) && count($args) <= $maxLen;
                if ($match)
                    $methods[] = $m;
            }
        
        if (count($methods) === 0)
            throw new \OneLang\Core\Error("Method '" . $methodName . "' was not found on type '" . $cls->name . "' with " . count($args) . " arguments");
        else if (count($methods) > 1) {
            // TODO: actually we should implement proper method shadowing here...
            $thisMethods = array_values(array_filter($methods, function ($x) use ($cls) { return $x->parentInterface === $cls; }));
            if (count($thisMethods) === 1)
                return $thisMethods[0];
            throw new \OneLang\Core\Error("Multiple methods found with name '" . $methodName . "' and " . count($args) . " arguments on type '" . $cls->name . "'");
        }
        return $methods[0];
    }
    
    protected function resolveReturnType($expr, $genericsResolver) {
        $genericsResolver->collectFromMethodCall($expr);
        
        for ($i = 0; $i < count($expr->args); $i++) {
            // actually doesn't have to resolve, but must check if generic type confirm the previous argument with the same generic type
            $paramType = $genericsResolver->resolveType($expr->method->parameters[$i]->type, false);
            if ($paramType !== null)
                $expr->args[$i]->setExpectedType($paramType);
            $expr->args[$i] = $this->main->runPluginsOn($expr->args[$i]);
            $genericsResolver->collectResolutionsFromActualType($paramType, $expr->args[$i]->actualType);
        }
        
        if ($expr->method->returns === null) {
            $this->errorMan->throw("Method (" . $expr->method->parentInterface->name . "::" . $expr->method->name . ") return type was not specified or infered before the call.");
            return;
        }
        
        $expr->setActualType($genericsResolver->resolveType($expr->method->returns, true), true, $expr instanceof InstanceMethodCallExpression && TypeHelper::isGeneric($expr->object->getType()));
    }
    
    protected function transformMethodCall($expr) {
        if ($expr->object instanceof ClassReference || $expr->object instanceof StaticThisReference) {
            $cls = $expr->object instanceof ClassReference ? $expr->object->decl : ($expr->object instanceof StaticThisReference ? $expr->object->cls : null);
            $method = $this->findMethod($cls, $expr->methodName, true, $expr->args);
            $result = new StaticMethodCallExpression($method, $expr->typeArgs, $expr->args, $expr->object instanceof StaticThisReference);
            $this->resolveReturnType($result, new GenericsResolver());
            return $result;
        }
        else {
            $resolvedObject = $expr->object->actualType !== null ? $expr->object : $this->main->runPluginsOn($expr->object);
            $objectType = $resolvedObject->getType();
            $intfType = $objectType instanceof ClassType ? $objectType->decl : ($objectType instanceof InterfaceType ? $objectType->decl : null);
            
            if ($intfType !== null) {
                $lambdaField = \OneLang\Core\ArrayHelper::find($intfType->fields, function ($x) use ($expr) { return $x->name === $expr->methodName && $x->type instanceof LambdaType && count($x->type->parameters) === count($expr->args); });
                if ($lambdaField !== null) {
                    $lambdaCall = new LambdaCallExpression(new InstanceFieldReference($expr->object, $lambdaField), $expr->args);
                    $lambdaCall->setActualType(($lambdaField->type)->returnType);
                    return $lambdaCall;
                }
                
                $method = $this->findMethod($intfType, $expr->methodName, false, $expr->args);
                $result = new InstanceMethodCallExpression($resolvedObject, $method, $expr->typeArgs, $expr->args);
                $this->resolveReturnType($result, GenericsResolver::fromObject($resolvedObject));
                return $result;
            }
            else if ($objectType instanceof AnyType) {
                $expr->setActualType(AnyType::$instance);
                return $expr;
            }
            else { }
            return $resolvedObject;
        }
    }
    
    function canTransform($expr) {
        return $expr instanceof UnresolvedMethodCallExpression && !($expr->actualType instanceof AnyType);
    }
    
    function transform($expr) {
        return $this->transformMethodCall($expr);
    }
}
